### PR TITLE
Add CJS export to conditional exports

### PR DESCRIPTION
**Pull request by Andarist**
Without this, it wouldn't be possible to `require('stylis')` in nodes supporting conditional exports as `require` cannot load modules (in contrary to `import` which can load commonjs scripts)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,9 +8,12 @@
   "repository": "https://github.com/thysultan/stylis.js",
   "bugs": "https://github.com/thysultan/stylis.js/issues",
   "type": "module",
-  "main": "dist/stylis.umd.js",
+  "main": "dist/umd/stylis.umd.js",
   "module": "dist/stylis.esm.js",
-  "exports": "./index.js",
+  "exports": {
+    "import": "./index.js",
```
Comment (Andarist) [inline package.json:14]: I personally don't quite like that a different file is being given to node (pkg.exports.import) and bundlers (pkg.module) - if the code is not specific to its environment.

Could we just point to the same file from both? I'm still not a fan of giving bundlers pre-minified code - had to debug few 3rd part dependencies in the past weeks and this would be a total pain if they would be minified (luckily, they weren't).

File: script/build.js [hunk 2]
```diff
@@ -1,7 +1,16 @@
 import {join} from 'path'
-import {terser} from "rollup-plugin-terser"
+import {terser} from 'rollup-plugin-terser'
 import size from 'rollup-plugin-size'
 
+const emitPackageScopeFile = (type) => {
```
Comment (Andarist) [inline script/build.js:5]: this is required so that commonjs/umd file can be treated as commonjs script while having `.js` extension (otherwise it gets treated as a module due to the closest package.json - the root one - having `"type":"module"`)

File: package.json [hunk 3]
```diff
@@ -8,9 +8,12 @@
   "repository": "https://github.com/thysultan/stylis.js",
   "bugs": "https://github.com/thysultan/stylis.js/issues",
   "type": "module",
-  "main": "dist/stylis.umd.js",
+  "main": "dist/umd/stylis.umd.js",
```
Comment (TrySound) [inline package.json:11]: Shouldn't commonjs modules have .cjs extension? I think this is a requirement when "type" is "module". The way around is package.json with "{ type: "commonjs" }" in subfolder.

Comment (Andarist) [reply package.json:11]: @TrySound - see a comment [here](https://github.com/thysultan/stylis.js/pull/207#discussion_r407931392) 😉

Comment (TrySound) [reply package.json:11]: Nice!

Comment (thysultan) [reply package.json:11]: Why the new directory?

Comment (TrySound) [reply package.json:11]: To add another package.json with { type: commonjs }

**Comment by thysultan** [comment]
>it wouldn't be possible to require('stylis')

Examples? As far as i can tell it should work without this as node versions that don't support modules will use the `pkg.main` field. It's unclear why we need this considering that.

**Comment by Andarist** [comment]
It's because in ESM-supporting node (v13) only `import` loader is capable of loading modules. This is documented [here](https://nodejs.org/api/esm.html#esm_conditional_exports):
> "require" - matched when the package is loaded via require(). As require() only supports CommonJS, the referenced file must be CommonJS.

I have also tested this myself before preparing this PR - can't require module, but can import a script (cjs) 🤷

**Comment by thysultan** [comment]
>only import loader is capable of loading modules

Yes that's the intended idea.

**Comment by Andarist** [comment]
Problem is that it means that **all dependents in the chain** must be modules and I'm afraid that the ecosystem is just not ready yet for this. In case of Emotion - nor React, nor Emotion distribute conditional exports just yet so while SSRing node will reach for their respective CJS files and try to access stylis on which it will fail because it's just a module without cjs fallback.

**Comment by thysultan** [comment]
Primary dependents should be bundling stylis and shipping that instead of including it as a direct dependency. This is just for people using stylis directly with "module" supported versions of NodeJS.

**Comment by Andarist** [comment]
Do you say that we at Emotion should bundle stylis? I wouldn't have expected that 😅 i much prefer not doing that as we would need to prepare a release after each new stylis release, but if that's your stance - sure, we can do that. What are the benefits of this approach though? Should this information be put in the docs?

An additional thing I don't quite like about this is that having stylis as a direct dependency would work when using node <=12 but suddenly break when upgrading to a newer one. From the consumer point of view, this is super unexpected.

File: package.json [hunk 4]
```diff
@@ -8,9 +8,12 @@
...
   "module": "dist/stylis.esm.js",
-  "exports": "./index.js",
+  "exports": {
+    "import": "./index.js",
+    "require": "./dist/umd/stylis.umd.js"
```
Comment (thysultan) [inline package.json:15]: Why can't this point directly to the file?

Comment (thysultan) [reply package.json:15]: By directly i mean directly to the current file `/dist/stylis.umd.js`

Comment (thysultan) [reply package.json:15]: For example in https://nodejs.org/api/esm.html#esm_approach_2_isolate_state they don't try to isolate the cjs file into a directory.

Comment (Andarist) [reply package.json:15]: > Why can't this point directly to the file? Doesn't the require imply cjs.

Unfortunately not, this is also kinda quirky for me and counter-intuitive. `"require"`/`"import"` keys only tell which loader should be used, but import is able to load cjs files, so it can't set the mode of the interpretation of the imported file. Maybe it's just a symmetry thing for `"require"`?

> For example in https://nodejs.org/api/esm.html#esm_approach_2_isolate_state they don't try to isolate the cjs file into a directory.

Yes, but they use `.cjs` extension - which sets the mode of the interpretation. `.mjs` files are always modules, `.cjs` are always scripts. I've just wanted to avoid `.cjs` extension here, especially given the fact that this is a UMD bundle which often is used when trying to use a particular lib directly in a browser, and `.cjs` doesn't make much sense in such context. We can, however, use `.cjs` to avoid this extra directory.

To sum it up:
- "require"/"import" - don't do anything in regards to module/script interpretation
- `.mjs`/`.cjs` extensions force a particular mode
- `.js` are parsed based on the nearest package.json and its "type", that's why we can create a "nested package scope", as the algorithm allows this, even though this section doesn't really mention this explicitly: https://nodejs.org/api/esm.html#esm_package_scope_and_file_extensions . This technique is used by some packages so a `.js` extension can be used for all files (for example Rollup does this - https://github.com/rollup/rollup/blob/c0c206e2a8a892fdb0e4f9b25b0a817b733a072f/rollup.config.js#L118 )

Comment (thysultan) [reply package.json:15]: >We can, however, use .cjs to avoid this extra directory.

If we can avoid this, let us export a .cjs file, i think rollup can produce a dedicated cjs file without the wrappers involved in the umd file.

Comment (Andarist) [reply package.json]: Ok, added a separate config item for `.cjs` file

File: script/build.js [hunk 5]
```diff
@@ -1,8 +1,8 @@
 import {join} from 'path'
-import {terser} from "rollup-plugin-terser"
+import {terser} from 'rollup-plugin-terser'
 import size from 'rollup-plugin-size'
 
-const options = {mangle: true, compress: false}
+const options = {mangle: true, compress: false, toplevel: true}
```
Comment (thysultan) [inline script/build.js:5]: What does the toplevel option do?

Comment (thysultan) [reply script/build.js:5]: >--toplevel                  Compress and/or mangle variables in top level scope.

NVM

**Comment by thysultan** [comment]
LGTM

File: package.json (hunk 1 above)
Comment (thysultan) [reply package.json:14]: We still want to point main to the umd file.

Comment (Andarist) [reply package.json:14]: Just because conceptually I see `pkg.main` as `pkg.exports.require`, so I've reused the same file - it also avoids UMD wrapper for node@<13 consumers, not a big deal - can change it back if you want.
